docs/DesignDocs/Attestation_API_Proposal.md

```diff
@@ -0,0 +1,534 @@
+- Unregister a verifier plugin.
+  - Function `oe_unregister_verifier(oe_verifier_t* plugin)`,
+  declared in header file [openenclave/attestation/plugin.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/plugin.h).
+
+The single supported SGX evidence format UUID, OE_SGX_PLUGIN_UUID for SGX
+ECDSA format, is declared in header file
+[openenclave/internal/sgx/plugin.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/internal/sgx/plugin.h).
+
+In addition to the above plugin-based public API for attestation, the OE SDK
+also supports legacy API for generation and verification of SGX reports:
+`oe_get_report()` and `oe_verify_report()`.
+
+Areas of improvement in the existing attestation public API:
+- Attester and Verifier plugins are unnecessarily exposed to application
+software
+  - It's sufficient for application to use globally unique evidence format
+  UUID in generation and verification of evidences.
+  - Plugins are the artifact of the OE SDK attestation framework internal
+  design that application software does not need to care.
+- The API does not enable applications to discover which formats are supported
+for evidence generation or verification.
+- SGX specific functions are exposed as public API
+- The public and plugin API declarations are all contained in a
+single header file [openenclave/attestation/plugin.h](https://github.com/openenclave/openenclave/blob/v0.9.x/include/openenclave/attestation/plugin.h).
+- For SGX, the globally unique UUIDs identify the evidence formats.
+There is no need of local / remote flag.
+
+## Proposed OE SDK V0.10 Public API
+
+Objectives: address the areas of improvement in the existing public API.
+
+- Separate public and plugin API declarations into different header files.
+- Only expose evidence formats for evidence generation and verification.
+There is no exposure of plugin design.
+- Remove TEE specific API elements, e.g. SGX-specific functions
+- Enable applications run-time discovery of supported formats for evidence
+generation and verification in
+[TEE-agnostic](https://github.com/shnwc/openenclave/blob/master/docs/DesignDocs/NotesOnAttestationAPI.md#attester--verifier-security-model-and-tee-agnostic-design)
+manner.
+- Remove legacy artifacts that have become redundant, e.g. the SGX
+local / remote flags.
+
+Proposed Headers files:
+
+- Attester public API to be declared in `openenclave/attestation/attesters.h`
+- Verifier public API to be declared in `openenclave/attestation/verifiers.h`
+- SGX evidence format UUIDs to be declared in
+`openenclave/attestation/evidence_uuids_sgx.h`
+- Plugin API (helper functions and plugin entry points) to be declared in
+`openenclave/internal/plugin.h`
+
+Use cases for evidence generation:
+
+- Initialize attester environment.
+  - Function `oe_result_t oe_attester_initialize(void)`.
```

Comment:
From application developers' point of view, this function sets a requirement for an attester application to declare its intention to start using the plugin-based attester APIs.

Internally, the OE SDK framework / plugin implementation of this API function can do anything as needed to get the environment  set up properly. If resolution of issue #2903 leads to a good plugin design that performs all initialization internally, at that time this function can become NOP.

@yentsanglee on your comment "Such initialization function does not work for external plugins.", I'd like to learn about the cases you see that external plugins can'be be initialized by some sort of implementation of this function.

Legacy function `oe_get_report()` could be implemented by the OE SDK framework to call `oe_attester_initializat()` first, so applications don't need to call `oe_attester_initialize()` explicitly. Since `oe_attester_initialize()` is idempotent, there is no harm calling it multiple times.

Reply:
For v0.10, we will proceed with this proposal. A longer term discussion around plugins needs to happen in SIG-Arch.